File: backend/app/core/contact_finder.py
"""Find relevant contacts at discovered companies."""

from __future__ import annotations

import re
import time
import requests
from bs4 import BeautifulSoup
from ddgs import DDGS
from config import TARGET_ROLES


HEADERS = {
    "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36"
}

# Common team/about page paths
TEAM_PATHS = [
    "/about", "/about-us", "/team", "/our-team", "/people",
    "/about/team", "/company", "/company/team", "/leadership",
]


def _fetch_page(url: str, timeout: int = 10) -> str | None:
    """Fetch a page and return its HTML, or None on failure."""
    try:
        resp = requests.get(url, headers=HEADERS, timeout=timeout, allow_redirects=True)
        if resp.status_code == 200:
            return resp.text
    except Exception:
        pass
    return None


def _extract_people_from_html(html: str, company_name: str) -> list[dict]:
    """
    Try to extract names and titles from a team/about page.
    Looks for common patterns in HTML structure.
    """
    soup = BeautifulSoup(html, "html.parser")
    people = []

    # Strategy 1: Look for structured elements with person info
    # Many team pages use cards with name in h2/h3/h4 and title in p/span
    for card in soup.find_all(["div", "li", "article"], class_=re.compile(
        r"team|member|person|staff|leader|bio|card", re.I
    )):
        name_el = card.find(["h2", "h3", "h4", "strong", "b"])
        if not name_el:
            continue
        name = name_el.get_text(strip=True)

        # Skip if doesn't look like a name (2-4 words, each capitalized)
        name_parts = name.split()
        if len(name_parts) < 2 or len(name_parts) > 5:
            continue
        if not all(p[0].isupper() for p in name_parts if p):
            continue

        # Find title nearby
        title = ""
        for sibling in [card.find("p"), card.find("span"), card.find(["h5", "h6"])]:
            if sibling:
                t = sibling.get_text(strip=True)
                if any(role in t.lower() for role in TARGET_ROLES) or len(t) < 60:
                    title = t
                    break

        if name and len(name) > 3:
            people.append({
                "name": name,
                "title": title or "Unknown",
                "company": company_name,
                "linkedin_url": "",
                "source": "team_page",
            })

    # Strategy 2: Look for LinkedIn links on the page
    for a in soup.find_all("a", href=re.compile(r"linkedin\.com/in/", re.I)):
        href = a["href"]
        # Try to get the name from link text or nearby elements
        link_text = a.get_text(strip=True)
        if link_text and len(link_text.split()) >= 2:
            # Check if we already have this person
            if not any(p["name"] == link_text for p in people):
                people.append({
                    "name": link_text,
                    "title": "Unknown",
                    "company": company_name,
                    "linkedin_url": href,
                    "source": "team_page",
                })

    return people


def _search_linkedin_contacts(company_name: str, domain: str) -> list[dict]:
    """Use DuckDuckGo to find LinkedIn profiles for people at a company."""
    people = []
    query = f'site:linkedin.com/in "{company_name}" founder OR CEO OR CTO OR "head of"'

    try:
        results = list(DDGS().text(query, max_results=5))
    except Exception:
        return people

    company_lower = company_name.lower()

    for r in results:
        url = r.get("href", "")
        title = r.get("title", "")
        body = r.get("body", "")

        if "linkedin.com/in/" not in url:
            continue

        # LinkedIn titles are usually "Name - Title - Company | LinkedIn"
        parts = title.replace(" | LinkedIn", "").split(" - ")
        name = parts[0].strip() if parts else ""
        role = parts[1].strip() if len(parts) > 1 else "Unknown"

        # Verify the company name actually appears in the result text
        # This prevents false matches (e.g., Jensen Huang showing up for Iodine Software)
        result_text = (title + " " + body).lower()
        if company_lower not in result_text and domain.split(".")[0] not in result_text:
            continue

        if name and len(name.split()) >= 2:
            people.append({
                "name": name,
                "title": role,
                "company": company_name,
                "linkedin_url": url,
                "source": "linkedin_search",
            })

    return people


def _is_valid_person_name(name: str) -> bool:
    """Check if a string looks like a real person's name."""
    if not name or len(name) < 4:
        return False
    parts = name.split()
    if len(parts) < 2 or len(parts) > 4:
        return False
    # Each part should be capitalized and alphabetic (allow hyphens, accents)
    for p in parts:
        if not p[0].isupper():
            return False
        # Skip if it contains numbers or looks like a product/company name
        if any(c.isdigit() for c in p):
            return False
    # Skip common non-name patterns
    lower = name.lower()
    skip = [
        "general partner", "managing director", "our ", "the ",
        "high-yield", "business", "savings", "parking", "conference",
        "award", "leadership", "client", "north", "waystar",
        "passport", "true north", "learn more", "about us",
        "atlas", "company", "innovative", "fast company",
        "thread mind", "san francisco", "austin", "new york",
        "most innovative", "venture partner", "special advisor",
        "operations analyst", "bangalore", "india", "chief executive",
        "officer", "director", "manager", "analyst", "engineer",
        "intern", "associate", "partner",
    ]
    if any(s in lower for s in skip):
        return False
    # Name parts shouldn't end with periods or contain special chars
    if "." in name or "'" in name or "™" in name:
        return False
    # Reject if name looks like it has a title concatenated (e.g. "Eric PoirierChief")
    for part in parts:
        if len(part) > 15:  # Real name parts are rarely this long
            return False
    return True


def find_contacts(companies: list[dict], max_contacts_per_company: int = 3) -> list[dict]:
    """
    For each company, find relevant contacts via team pages and LinkedIn search.
    """
    all_contacts = []
    seen_keys = set()  # (name_lower, domain) for dedup

    for i, company in enumerate(companies):
        name = company["name"]
        domain = company["domain"]
        website = company["website"]

        print(f"  [{i+1}/{len(companies)}] Finding contacts at {name}...")

        contacts_for_company = []

        # Method 1: Scrape team/about pages
        for path in TEAM_PATHS:
            html = _fetch_page(f"{website}{path}")
            if html:
                people = _extract_people_from_html(html, name)
                contacts_for_company.extend(people)
                if people:
                    break

        # Method 2: LinkedIn search (if we didn't find enough from team page)
        if len(contacts_for_company) < 2:
            linkedin_people = _search_linkedin_contacts(name, domain)
            existing_names = {c["name"].lower() for c in contacts_for_company}
            for p in linkedin_people:
                if p["name"].lower() not in existing_names:
                    contacts_for_company.append(p)

        # Filter: must be a valid person name
        contacts_for_company = [c for c in contacts_for_company if _is_valid_person_name(c["name"])]

        # Filter to relevant roles and cap per company
        relevant = []
        other = []
        for c in contacts_for_company:
            title_lower = c["title"].lower()
            if any(role in title_lower for role in TARGET_ROLES):
                relevant.append(c)
            else:
                other.append(c)

        final = (relevant + other)[:max_contacts_per_company]

        # Attach company metadata and deduplicate
        for c in final:
            key = (c["name"].lower(), domain)
            if key in seen_keys:
                continue
            seen_keys.add(key)

            c["company_website"] = website
            c["company_domain"] = domain
            c["company_location"] = company["location"]
            c["company_industry"] = ", ".join(company["industry_tags"])
            c["description"] = company.get("description", "")

            all_contacts.append(c)

        # Rate limiting
        if i < len(companies) - 1:
            time.sleep(1)

    print(f"\n  Found {len(all_contacts)} total contacts")
    return all_contacts


if __name__ == "__main__":
    # Quick test with a couple companies
    test_companies = [
        {"name": "Anthropic", "domain": "anthropic.com", "website": "https://anthropic.com",
         "location": "Remote", "industry_tags": ["AI"]},
    ]
    contacts = find_contacts(test_companies)
    for c in contacts:
        print(f"  {c['name']} — {c['title']} at {c['company']}")

File: src/lib/discover/contact-finder.ts
// Contact extractor — Stage 3 of the discover pipeline.
//
// Ported from backend/app/core/contact_finder.py. For each candidate
// company, walks the well-known team / about / leadership page paths and
// scrapes person cards with cheerio. Falls back to a DDG LinkedIn dork
// (`site:linkedin.com/in "Company Name" analyst`) when the team page
// returns nothing useful. Every extracted name flows through
// isValidPersonName() before it can become a ContactCandidate.
//
// Identity anchoring: every candidate carries a `sourceUrl` that points
// to the page the name was actually found on (the team page or the
// LinkedIn search-result snippet). Stage 4 (signal-detector) refuses to
// attach signals to a contact unless the signal source mentions the
// contact's name in the same document.

import * as cheerio from "cheerio";
import { isValidPersonName } from "./name-validator";
import { decodeDdgHref } from "./entity-finder";

const USER_AGENT =
  "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36";

const DDG_HTML_URL = "https://html.duckduckgo.com/html/";

// Common path conventions for company "team / about / leadership" pages.
// Lifted from contact_finder.py:TEAM_PATHS, ordered most-likely-first
// so we exit early on a hit.
export const TEAM_PATHS: readonly string[] = [
  "/team",
  "/about/team",
  "/our-team",
  "/people",
  "/leadership",
  "/about",
  "/about-us",
  "/company/team",
  "/company",
];

// Roles that mark a person as worth surfacing for student warm outreach.
// Merges contact_finder.py TARGET_ROLES (founder/CEO-heavy, geared for
// tech recruiters) with the IB-specific roles students at KithNode are
// actually trying to reach (analyst, associate, MD, partner).
//
// Matching is word-boundary regex (not substring) so "coo" doesn't match
// "coordinator" and "md" doesn't match "Madison".
export const TARGET_ROLE_PATTERN: RegExp = new RegExp(
  "\\b(?:" +
    [
      "analysts?", "associates?", "vice presidents?", "vps?",
      "managing directors?", "mds?", "principals?", "partners?", "directors?",
      "founders?", "co-?founders?", "ceos?", "ctos?", "coos?", "cfos?",
      "presidents?", "head of", "hiring managers?", "recruiters?",
      "talent", "people operations",
    ].join("|") +
    ")\\b",
  "i",
);

export interface ContactCandidate {
  name: string;
  title: string;
  company: string;
  companyDomain: string;
  linkedinUrl: string;
  source: "team_page" | "linkedin_search";
  /** The page where this person was actually found. Used by Stage 4
   *  to enforce identity anchoring on every signal. */
  sourceUrl: string;
}

export interface CompanyInput {
  name: string;
  domain: string;
  /** Base URL — `https://${domain}` works as a default. */
  website: string;
}

export interface FindContactsOptions {
  /** Cap per company to avoid runaway scraping on huge team pages. */
  maxPerCompany?: number;
  /** Throttle between companies. */
  throttleMs?: number;
  /** Skip the LinkedIn dork fallback (useful for tests / quick mode). */
  skipLinkedInFallback?: boolean;
  /** Called once per company so the caller can emit progress events. */
  onProgress?: (info: {
    index: number;
    total: number;
    company: CompanyInput;
    foundForCompany: number;
    foundTotal: number;
  }) => void | Promise<void>;
  /** Abort signal — checked between companies to exit early on cancel. */
  signal?: AbortSignal;
}

// ── Page fetching ─────────────────────────────────────────────────────

export async function fetchPage(url: string): Promise<string | null> {
  try {
    const res = await fetch(url, {
      headers: { "User-Agent": USER_AGENT, Accept: "text/html" },
      redirect: "follow",
    });
    if (!res.ok) return null;
    return await res.text();
  } catch {
    return null;
  }
}

// ── Team page extraction ──────────────────────────────────────────────

const TEAM_CARD_SELECTOR = "div, li, article";
const TEAM_CARD_CLASS_RE = /team|member|person|staff|leader|bio|card/i;

/**
 * Extract person cards from a team / about page. Mirrors the cheerio
 * equivalent of contact_finder.py:_extract_people_from_html(). Looks for
 * containers whose class name suggests a person card, then pulls the
 * name from the first heading-like child and the title from a nearby
 * paragraph or span.
 */
export function extractPeopleFromHtml(
  html: string,
  company: string,
  companyDomain: string,
  sourceUrl: string,
): ContactCandidate[] {
  const $ = cheerio.load(html);
  const out: ContactCandidate[] = [];
  const seen = new Set<string>();

  $(TEAM_CARD_SELECTOR).each((_, el) => {
    const $card = $(el);
    const className = $card.attr("class") || "";
    if (!TEAM_CARD_CLASS_RE.test(className)) return;

    const nameEl = $card.find("h2, h3, h4, strong, b").first();
    if (!nameEl.length) return;
    const name = nameEl.text().trim();
    if (!isValidPersonName(name)) return;

    // Title: first p / span / h5 / h6 nearby. Cap length so we don't
    // grab a whole bio paragraph.
    let title = "Unknown";
    const titleCandidates = [
      $card.find("p").first(),
      $card.find("span").first(),
      $card.find("h5, h6").first(),
    ];
    for (const candidate of titleCandidates) {
      if (!candidate.length) continue;
      const t = candidate.text().trim();
      if (!t) continue;
      const matchesRole = TARGET_ROLE_PATTERN.test(t);
      if (matchesRole || t.length < 60) {
        title = t;
        break;
      }
    }

    // Optional LinkedIn link inside the same card
    let linkedinUrl = "";
    const liLink = $card.find("a[href*='linkedin.com/in/']").first();
    if (liLink.length) linkedinUrl = liLink.attr("href") || "";

    const key = name.toLowerCase();
    if (seen.has(key)) return;
    seen.add(key);

    out.push({
      name,
      title,
      company,
      companyDomain,
      linkedinUrl,
      source: "team_page",
      sourceUrl,
    });
  });

  // Strategy 2: any anchor pointing at a LinkedIn profile, with the link
  // text as the name. Catches cards we couldn't class-match above.
  $("a[href*='linkedin.com/in/']").each((_, a) => {
    const $a = $(a);
    const href = $a.attr("href") || "";
    const text = $a.text().trim();
    if (!text || !isValidPersonName(text)) return;
    const key = text.toLowerCase();
    if (seen.has(key)) return;
    seen.add(key);
    out.push({
      name: text,
      title: "Unknown",
      company,
      companyDomain,
      linkedinUrl: href,
      source: "team_page",
      sourceUrl,
    });
  });

  return out;
}

// ── DDG LinkedIn dork fallback ────────────────────────────────────────

/**
 * Last-resort person discovery: a `site:linkedin.com/in` Google-dork
 * against DDG. We deliberately bypass entity-finder.ts's blocklist here
 * because LinkedIn is exactly what we want this time.
 *
 * Note: requires identity anchoring downstream. The company name MUST
 * appear in the result snippet, otherwise we drop the row — this is the
 * fix for the Python bot's v1 false-attribution bug.
 */
export async function searchLinkedInContacts(
  companyName: string,
  companyDomain: string,
): Promise<ContactCandidate[]> {
  const query = `site:linkedin.com/in "${companyName}" analyst OR associate OR founder OR CEO OR "managing director"`;
  const params = new URLSearchParams({ q: query });
  let html: string;
  try {
    const res = await fetch(`${DDG_HTML_URL}?${params}`, {
      headers: { "User-Agent": USER_AGENT, Accept: "text/html" },
      redirect: "follow",
    });
    if (!res.ok) return [];
    html = await res.text();
  } catch {
    return [];
  }

  const $ = cheerio.load(html);
  const out: ContactCandidate[] = [];
  const seen = new Set<string>();
  const companyLower = companyName.toLowerCase();
  const domainStem = companyDomain.split(".")[0]?.toLowerCase() || "";

  $(".result").each((_, el) => {
    const $el = $(el);
    const linkEl = $el.find("a.result__a").first();
    const titleText = linkEl.text().trim();
    const url = decodeDdgHref(linkEl.attr("href") || "");
    if (!url.includes("linkedin.com/in/")) return;

    // LinkedIn result titles look like: "Name - Title - Company | LinkedIn"
    const cleaned = titleText.replace(" | LinkedIn", "").trim();
    const parts = cleaned.split(" - ").map((p) => p.trim());
    const name = parts[0] || "";
    const role = parts[1] || "Unknown";

    // Identity anchoring: the company name must appear in the LinkedIn
    // title's company segment(s) — not anywhere in the snippet. The
    // snippet is summary text and frequently says "...has nothing to do
    // with Acme..." which would false-positive a substring check.
    const companySegments = parts.slice(2).map((p) => p.toLowerCase()).join(" ");
    const anchored =
      companySegments.includes(companyLower) ||
      (!!domainStem && companySegments.includes(domainStem));
    if (!anchored) return;

    if (!isValidPersonName(name)) return;
    const key = name.toLowerCase();
    if (seen.has(key)) return;
    seen.add(key);

    out.push({
      name,
      title: role,
      company: companyName,
      companyDomain,
      linkedinUrl: url,
      source: "linkedin_search",
      sourceUrl: url,
    });
  });

  return out;
}

// ── Orchestration ─────────────────────────────────────────────────────

export async function findContacts(
  companies: readonly CompanyInput[],
  options: FindContactsOptions = {},
): Promise<ContactCandidate[]> {
  const {
    maxPerCompany = 3,
    throttleMs = 1000,
    skipLinkedInFallback = false,
    onProgress,
    signal,
  } = options;
  const all: ContactCandidate[] = [];
  const seenGlobal = new Set<string>();

  for (let i = 0; i < companies.length; i++) {
    if (signal?.aborted) break;
    const company = companies[i];
    const found: ContactCandidate[] = [];

    // Try team / about / leadership pages first.
    for (const path of TEAM_PATHS) {
      const url = `${company.website.replace(/\/$/, "")}${path}`;
      const html = await fetchPage(url);
      if (!html) continue;
      const people = extractPeopleFromHtml(html, company.name, company.domain, url);
      if (people.length) {
        found.push(...people);
        break;
      }
    }

    // Fallback: DDG LinkedIn dork — only if team scrape gave us <2.
    if (found.length < 2 && !skipLinkedInFallback) {
      const linkedinPeople = await searchLinkedInContacts(company.name, company.domain);
      const localSeen = new Set(found.map((p) => p.name.toLowerCase()));
      for (const p of linkedinPeople) {
        if (!localSeen.has(p.name.toLowerCase())) found.push(p);
      }
    }

    // Filter to roles we care about, then take top N.
    const ranked = rankByRole(found);
    let acceptedForCompany = 0;
    for (const c of ranked.slice(0, maxPerCompany)) {
      const key = `${c.name.toLowerCase()}|${company.domain}`;
      if (seenGlobal.has(key)) continue;
      seenGlobal.add(key);
      all.push(c);
      acceptedForCompany++;
    }

    if (onProgress) {
      await onProgress({
        index: i,
        total: companies.length,
        company,
        foundForCompany: acceptedForCompany,
        foundTotal: all.length,
      });
    }

    if (i < companies.length - 1 && throttleMs > 0) {
      await new Promise((resolve) => setTimeout(resolve, throttleMs));
    }
  }

  return all;
}

/**
 * Re-orders candidates so that ones with a TARGET_ROLES title float to
 * the top, preserving the original order within each bucket. Lets the
 * `maxPerCompany` slice keep the relevant rows even on a dense team page.
 */
export function rankByRole(candidates: ContactCandidate[]): ContactCandidate[] {
  const relevant: ContactCandidate[] = [];
  const other: ContactCandidate[] = [];
  for (const c of candidates) {
    if (TARGET_ROLE_PATTERN.test(c.title)) relevant.push(c);
    else other.push(c);
  }
  return [...relevant, ...other];
}
